pkg/kubelet/images/image_manager.go

```diff
@@ -64,27 +79,44 @@ func NewImageManager(recorder record.EventRecorder, imageService kubecontainer.I
 	} else {
 		puller = newParallelImagePuller(imageService, maxParallelImagePulls)
 	}
+
 	return &imageManager{
-		recorder:               recorder,
-		imageService:           imageService,
-		backOff:                imageBackOff,
-		puller:                 puller,
-		podPullingTimeRecorder: podPullingTimeRecorder,
+		recorder:                     recorder,
+		imageService:                 imageService,
+		backOff:                      imageBackOff,
+		puller:                       puller,
+		podPullingTimeRecorder:       podPullingTimeRecorder,
+		keyring:                      keyring,
+		ensureSecretPulledImage:      make(map[string]*ensureSecretPulledImageDigest),
+		pullImageSecretRecheckPeriod: pullImageSecretRecheckPeriod,
 	}
 }
 
 // shouldPullImage returns whether we should pull an image according to
 // the presence and pull policy of the image.
-func shouldPullImage(container *v1.Container, imagePresent bool) bool {
-	if container.ImagePullPolicy == v1.PullNever {
+func shouldPullImage(container *v1.Container, imagePresent, pulledBySecret, ensuredBySecret bool) bool {
+	switch container.ImagePullPolicy {
+	case v1.PullNever:
 		return false
-	}
-
-	if container.ImagePullPolicy == v1.PullAlways ||
-		(container.ImagePullPolicy == v1.PullIfNotPresent && (!imagePresent)) {
+	case v1.PullAlways:
 		return true
+	case v1.PullIfNotPresent:
+		if !utilfeature.DefaultFeatureGate.Enabled(features.KubeletEnsureSecretPulledImages) {
```

Comment:
nit: This may read easier: 
```
	case v1.PullIfNotPresent:
		// pull if image doesn't exist
		if !imagePresent {
			return true
		}
		// if the imageRef has been pulled by a secret and Pull Policy is PullIfNotPresent
		// we need to ensure that the current pod's secrets map to an auth that has Already
		// pulled the image successfully. Otherwise pod B could use pod A's images
		// without auth. So in this case if pulledBySecret but not ensured by matching
		// secret auth for a pull again for the pod B scenario where the auth does not match
		if utilfeature.DefaultFeatureGate.Enabled((features.KubeletEnsureSecretPulledImages)) {
			if pulledBySecret && !ensuredBySecret {
				return true // noting here that old behaviour returns false in this case indicating the image should not be pulled
			}
		}
		return false
```

Reply:
agreed that reads cleaner .. because the feature gate only applies to present images..